### PR TITLE
DOC: update `stats`, `integrate`, `optimize`, and `differentiate` sections of roadmap

### PR DESCRIPTION
**Pull request by mdhaber**
#### What does this implement/fix?
Aside from `scipy.sparse`, it looks like the last major updates to the roadmap was in 2022. I had some thoughts on where `stats`, `integrate`, `optimize`, and `differentiate` might be headed, so I thought I'd propose those here.

### CODE REVIEW

File: doc/source/dev/roadmap-detailed.rst [hunk 1]
```diff
@@ -78,8 +78,8 @@ Pythran is still an optional build dependency, and can be disabled with
 happen it must be clear that the maintenance burden is low enough.
 
 
-Use of venerable Fortran libraries
```
Comment (mdhaber) [inline doc/source/dev/roadmap-detailed.rst:81]: The trigram "venerable Fortran library" stands out to me in SciPy documentation. I thought "venerable" might not be needed every time we mention Fortran libraries.

File: doc/source/dev/roadmap-detailed.rst [hunk 2]
```diff
@@ -123,8 +123,26 @@ feature requests.
...
+This module needs only periodic updates to the numerical values.
 
+differentiate
+`````````````
+This module was added with the understanding that its scope would be limited.
+The goal is to provide basic differentiation of black-box functions, not
+replicate all features of existing differentiation tools. With that in mind,
+items for future work include:
+
+- Improve support for functions that accept additional arguments (e.g. add
```
Comment (mdhaber) [inline doc/source/dev/roadmap-detailed.rst:135]:
```suggestion
- Improve support for callables that accept additional arguments (e.g. add
```

File: doc/source/dev/roadmap-detailed.rst [hunk 3]
```diff
@@ -123,8 +123,26 @@ feature requests.
...
+  ``*args`` to ``jacobian`` and ``hessian``). Note that this is not trivial
+  due to the way elements of the arrays are eliminated when their corresponding
+  calculations have converged.
+- Improve implementation of `scipy.differentiate.hessian`: rather than chaining
+  first-order differentiation, use a second-order differentiation stencil.
+- Consider the addition of an option to use relative step sizes.
+- Consider generalizing the approach to use "polynomial extrapolation"; i.e.,
```
Comment (mdhaber) [inline doc/source/dev/roadmap-detailed.rst:142]: "polynomial extrapolation" was the term used in gh-17059 for this idea, AFAICT. Regardless of what the implementation *looks* like or how the theory is presented (Richardson extrapolation, etc...) , I think finite step numerical differentiation is equivalent to fitting a polynomial to the data and differentiating the polynomial exactly. A key difference between [Jacobi](https://hdembinski.github.io/jacobi/) from what we currently do is that our method is equivalent to exactly fitting a polynomial of order $n$ with $n+1$ points whereas Jacobi uses `polyfit` to fit a polynomial of order $n$ with $n + m$. This is discussed a bit in mdhaber/jacobi#1,

File: doc/source/dev/roadmap-detailed.rst [hunk 4]
```diff
@@ -272,23 +288,16 @@ maintenance.  No major plans or wishes here.
...
-  however; we need to make it faster and make it easier to compare performance of
-  optimizers via plotting performance profiles.
+It is very challenging to write and maintain optimizers that approach the
+state of the art for large-scale problems. In many cases, it will be more
+appropriate for SciPy to add interfaces to carefully-vetted,
+externally-maintained projects (e.g. HiGHS, PRIMA) rather than vendor
+existing optimizers or introduce custom implementations. Other areas
+for future work include the following.
 
+- Improve the interfaces of existing optimizers (e.g. ``shgo``).
```
Comment (mdhaber) [inline doc/source/dev/roadmap-detailed.rst:298]:
```suggestion
- Improve the interfaces of existing optimizers (e.g. ``shgo``).
```
I single out SHGO mostly to give an idea of what I mean - I think many of the [issues](https://github.com/scipy/scipy/issues?q=is%3Aissue+is%3Aopen+shgo+label%3Ascipy.optimize) have to do with a mismatch between the documented options and what the code does. Other functions have a ton of options, and I'm not sure that the interfaces are as consistent as they could be or that all the options are well documented/tested.

File: doc/source/dev/roadmap-detailed.rst [hunk 5]
```diff
@@ -157,8 +173,8 @@ includes
       algorithms by Dierckx, de Boor, possibly other). 
 
 Once we have a reasonably feature complete set, we can start taking a long look
-at the future of the venerable FITPACK Fortran library, which currently is the
-only way of constructing smoothing splines in SciPy.
+at the future of the FITPACK Fortran library, which currently is the only way
+of constructing smoothing splines in SciPy.
```
Comment (mdhaber) [inline doc/source/dev/roadmap-detailed.rst:177]: @ev-br et al, I would really love to see splines support for a) n-d `x` and b) other array API compatible backends. Is that in the cards?

File: doc/source/dev/roadmap-detailed.rst [hunk 6]
```diff
@@ -133,15 +151,13 @@ This module is in good shape.
...
-  In the future we can consider (soft-)deprecating the older API.
-
-The numerical integration functions are in good shape.  Support for integrating
-complex-valued functions and integrating multiple intervals (see `gh-3325
-<https://github.com/scipy/scipy/issues/3325>`__) could be added.
+- Complete the feature set of the new ``cubature`` function, and add an interface
+  tailored to one-dimensional integrals.
+- Migrate functions for generating quadrature rule points and weights from `special`,
+  improve their reliability, and add support for other important rules.
+- Complete the feature set of ``solve_ivp``, including all functionality of the
```
Comment (j-bowhay) [inline doc/source/dev/roadmap-detailed.rst:158]: A couple of possible improvements:
- The methods we have are quite outdated eg. `rk45` should be replaced with `tsit5` see https://docs.sciml.ai/DiffEqDocs/stable/solvers/ode_solve/#Translations-from-MATLAB/Python/R
- Investigate fast/low precision exponentiation for the adaptive step size calculation 
- Investigate changing the step size control from an integral controller to a pi controller, see https://docs.sciml.ai/DiffEqDocs/stable/extras/timestepping/
- Accept low level callables
See https://www.stochasticlifestyle.com/comparison-differential-equation-solver-suites-matlab-r-julia-python-c-fortran/ for more areas of improvement

Comment (j-bowhay) [reply doc/source/dev/roadmap-detailed.rst:158]: It would also be great if SciPy had a dae solver

File: doc/source/dev/roadmap-detailed.rst [hunk 7]
```diff
@@ -272,23 +288,16 @@ maintenance.  No major plans or wishes here.
...
-  (e.g. with summary plots).
-- deprecate the ``fmin_*`` functions in the documentation, ``minimize`` is
-  preferred.
-- ``scipy.optimize`` has an extensive set of benchmarks for accuracy and speed of
-  the global optimizers. That has allowed adding new optimizers (``shgo`` and
-  ``dual_annealing``) with significantly better performance than the existing
-  ones.  The ``optimize`` benchmark system itself is slow and hard to use
-  however; we need to make it faster and make it easier to compare performance of
-  optimizers via plotting performance profiles.
+It is very challenging to write and maintain optimizers that approach the
```
Comment (j-bowhay) [inline doc/source/dev/roadmap-detailed.rst:291]: Agree, I think this the right direction

**Review by j-bowhay — COMMENTED** [review]
Looks good, I left some possible discussion points but they don't necessarily need to be included

File: doc/source/dev/roadmap-detailed.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:142]: The functions were originally developed to meet internal needs, and we are releasing them publicly because there has been demand for a while. We hope for the maintenance burden to be low if we keep the scope limited.  Of course, it's always up for discussion whether other methods should be added, but I phrased it like this to discourage developing features just for the sake of completness. I don't think we should try to compete with more specialized numerical differentiation packages.

> complex step

Perhaps there is space for it if there are not other vectorized, array API compatible implementations out there. Should they be contributed to other packages instead? Crossref gh-7195.

> quadrature based approaches

Are these actually distinct mathematically (that is, are net weights on each function value different?), or do they differ in implementation details?

File: doc/source/dev/roadmap-detailed.rst (hunk 6 above)
Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:158]: I wrote this part mostly from the perspective of letting go of the old interface. Personally, I'd be very interested in working on Array API compatible ODE/DAE solvers, but again I think there needs to be a study of what needs are not met by the rest of the ecosystem.

File: doc/source/dev/roadmap-detailed.rst (hunk 5 above)
Comment (ev-br) [reply doc/source/dev/roadmap-detailed.rst:177]: a) d>1 batches of the independent variable? Not sure TBH. Will complicate the interface at likely no perf gain. Would love to see a usage example where a python loop at call site does not cut it. Could you show one? Will be easier to play around with ergonomics. 

b) is definitely planned, yes! Effort most welcome. low level code is compiled though, so the GPU support will have to be limited to delegating (cupyx.scipy.interpolate is a thing). A technical thing to solve is how to do delegation for classes, not just functions.

Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:177]: > Will complicate the interface...

How so? Instead of requiring 1D `x` and broadcasting it (implicitly) to the shape of `y`, we would accept n-d `x` and broadcast it to the shape of `y`. Seems natural to me coming from `stats`. (e.g. any paired sample statistic used with `axis`.)

> Would love to see a usage example where a python loop at call site does not cut it

It's mostly that there may be many call sites and it's undesirable to have loops and calls to `xp.asarray` at all of them.

The application I have in mind is approximating probability distribution functions with splines like in `scipy.stats.sampling`/UNURAN, but built with SciPy primitives. The advantage to implementing (a simplified version of) these techniques ourselves is that it would be much faster to generate the interpolating splines with vectorized function calls. (The old `NumericalInverseHermite` was way faster to instantiate than the new one, and only marginally slower at generating extremely large samples.)

As for efficiency, suppose `searchsorted` were to support N-D input. With that, I can imagine implementing an efficient piecewise linear interpolator for N-D `x` and `y` using only array API calls. With a linear solver that supports batch input, could the same be done for higher order splines?

Comment (ev-br) [reply doc/source/dev/roadmap-detailed.rst:177]: The analogy with pairs sample statistics only takes us that far, same for searchsorted: for an interpolation problem there's an inherent asymmetry betwen `x` and `y`. 

However this starts getting too abstract, so how about we try a real example, to see how it goes and what the actual speed bumps are.
So, `stats.sampling` --- let's start from `NumericalInverseHermite`, is this a good starting point?

File: doc/source/dev/roadmap-detailed.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (ev-br) [reply doc/source/dev/roadmap-detailed.rst:142]: Complex step differentiation for hessians will suffer from roundoff, it seems:  https://www.acsu.buffalo.edu/~johnc/complex_step08.pdf

For 1st derivatives, `approx_derivative` implements it, https://github.com/scipy/scipy/blob/main/scipy/optimize/_numdiff.py#L612, and making it array api compatible looks straightforward.

File: doc/source/dev/roadmap-detailed.rst [hunk 8]
```diff
@@ -78,8 +78,8 @@ Pythran is still an optional build dependency, and can be disabled with
 happen it must be clear that the maintenance burden is low enough.
 
 
-Use of venerable Fortran libraries
-``````````````````````````````````
+Use of Fortran libraries
+````````````````````````
```
Comment (rgommers) [inline doc/source/dev/roadmap-detailed.rst:82]: If we're touching this anyway, let's update the text to say that the goal is to remove all Fortran code, and reference https://github.com/scipy/scipy/issues/18566. We're now far enough along and the end goal seems within reach, so time to make it official.

**Review by rgommers — COMMENTED** [review]
Overall this looks good to me, thanks for the update Matt!

File: doc/source/dev/roadmap-detailed.rst (hunk 6 above)
Comment (lucascolley) [reply doc/source/dev/roadmap-detailed.rst:158]: discussion on DAE in https://github.com/scipy/scipy/pull/13068

**Review by lucascolley — COMMENTED** [review]
RE: roadmap:

- [ ] The support for distributed and GPU arrays section is outdated. We should link to https://github.com/scipy/scipy/issues/18867 and https://github.com/scipy/scipy/issues/18286.
- [x] Is the transonic section still relevant @rgommers ?

RE: detailed roadmap:
- [ ] "Reliability of CI has not been good recently" seems outdated. Also the discussion of dropping distutils
- [x] extraneous backtick on line 111
- [ ] linalg: "Accelerate support is dropped because Apple doesn’t update Accelerate anymore"
- [x] misc: all functions have been removed now
- [ ] odr: https://github.com/scipy/scipy/issues/7107
- [ ] signal: is anything out of date here @DietBru ?
- [ ] sparse: can mention work on ND COO https://github.com/scipy/scipy/issues/21923
- [ ] special: should mention xsf and link to https://github.com/scipy/xsf/issues/1

**Comment by rgommers** [comment]
> Is the transonic section still relevant @rgommers ?

This bit can be removed: _"Enabling the use of an accelerator like Pythran, possibly via Transonic, and"_

The rest of the performance section still seems relevant.

File: doc/source/dev/roadmap-detailed.rst (hunk 6 above)
Comment (j-bowhay) [reply doc/source/dev/roadmap-detailed.rst:158]: > Looks like you have done some research - does SciPy need to step up in this department, or are other Python packages active and better places for the best new methods?
I try to keep an eye on what's happening in the Julia world, they have brought some really innovation to methods that haven't seen any innovation in a long time. 

It depends how flexible your requirements are:
- https://github.com/SciML/diffeqpy wraps differentialequations.jl and is great but requires you to have Julia install
- https://github.com/patrick-kidger/diffrax looks good although I have only looked at a surface level, only works for jax

File: doc/source/dev/roadmap-detailed.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (j-bowhay) [reply doc/source/dev/roadmap-detailed.rst:142]: > Are these distinct mathematically (that is, are net weights on each function value different?), or do they just differ in derivation or implementation details? For instance, I think I found that Richardson extrapolation was more numerically stable for high order than what I have, but I believe it would be equivalent (for a given stencil) with exact arithmetic. (IIUC, if the weights have to cancel out the low-order terms of the Taylor series, there's not really much flexibility if you want to get an n th order rule with n + 1 function evaluations.)

I agree the foundation of all methods is approximate the function (although it doesn't necessarily have to be with a polynomial eg. rational functions are better if there are poles) and then differentiate the approximation. However if you allow function evaluations on a disk in the complex plane then you can avoid the issues associated with standard finite differences, see J. N. Lyness and C. B. Moler, Numerical differentiation of analytic functions or NUMERICAL ALGORITHMS BASED ON ANALYTIC FUNCTION
 VALUES AT ROOTS OF UNITY,
 ANTHONY P. AUSTIN
, PETER KRAVANJA, AND LLOYD N. TREFETHEN. `mpmath`'s `diff` offers a method along these lines.
Note, I have no issue with the roadmap as it stands I just wanted to raise a discussion point:)

File: doc/source/dev/roadmap-detailed.rst [hunk 9]
```diff
@@ -272,23 +288,16 @@ maintenance.  No major plans or wishes here.
...
-  the global optimizers. That has allowed adding new optimizers (``shgo`` and
-  ``dual_annealing``) with significantly better performance than the existing
-  ones.  The ``optimize`` benchmark system itself is slow and hard to use
-  however; we need to make it faster and make it easier to compare performance of
-  optimizers via plotting performance profiles.
+It is very challenging to write and maintain optimizers that approach the
+state of the art for large-scale problems. In many cases, it will be more
+appropriate for SciPy to add interfaces to carefully-vetted,
+externally-maintained projects (e.g. HiGHS, PRIMA) rather than vendor
+existing optimizers or introduce custom implementations. Other areas
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:295]: Ok, I was trying to make a distinction between copy-pasting libraries with the idea of maintaining it ourselves vs providing an interface to a separately-maintained library. Is there a better word for the former?

File: doc/source/dev/roadmap-detailed.rst (hunk 8 above)
Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:82]: I can do that.

File: doc/source/dev/roadmap-detailed.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply doc/source/dev/roadmap-detailed.rst:295]: Not sure about a single word/phrase. I'd say something like: 

_"We aim to continuously improve the set of optimizers provided by this module. For small to medium scale problems, `scipy.optimize` is in fairly good shape. For large scale problems, the state of the art continues to advance and we aim to keep up by leveraging implementations from domain-specific libraries like HiGHS and PRIMA."_

File: doc/source/dev/roadmap-detailed.rst (hunk 8 above)
Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:82]: It was:

>  SciPy owes a lot of its success to relying on wrapping well established
Fortran libraries (QUADPACK, FITPACK, ODRPACK, ODEPACK etc). Some of these
libraries are aging well, others less so. We should audit our use of these
libraries with respect to the maintenance effort, the functionality, and the
existence of (possibly partial) alternatives, *including those inside SciPy*.

So replace all but the first sentence with something like

> For `<reasons>`, however, we are transitioning all Fortran libraries to more maintainable
implementations; see `gh-18566 <https://github.com/scipy/scipy/issues/18566>`__.

? 

Could you fill in `<reasons>`?

File: doc/source/dev/roadmap-detailed.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:142]: Right, that sounds like a generalization of the complex step? I meant for real steps.

**Comment by ev-br** [comment]
Now that we're updating the detailed roadmap, I'd rephrase the interpolate part along the lines of

FITPACK replacements
---------------------------------

We need to finish providing functional equivalents of the FITPACK Fortran library. The aim is to reimplement the algorithmic content of the FITPACK monolith in a modular fashion, to allow better user control and extensibility.
To this end, we need

- 1D splines: add the periodic spline fitting to `make_splrep` function.
- 2D splines: provide functional replacements of the `BivariateSpline` family of classes for scattered and gridded data.  

Once we have a feature-complete set of replacements, we can gracefully sunset our use of FITPACK.

Ideas for new features
---------------------------------

- add the ability to construct smoothing splines with variable number of knots to `make_smoothing_spline`. Currently, `make_smoothing_spline` always uses all data points for the knot vector.
- experiment with user-selectable smoothing criteria: add the P-splines penality form to complement the 2nd derivative penality of `make_smoothing_spline` and jumps of the 3rd derivative of `make_splrep`
- investigate ways of vectorizing the spline construction for batches of the independent variable. This class of features have been requested by `scipy.stats` developers
- allow specifying the boundary conditions for spline fitting.
- investigate constrained spline fitting problems which the FITPACK library was providing in Fortran, but were never exposed to the python interfaces
- NURBS support can be added if there is sufficient user interest

Scalability and performance
-----------------------------------------

We need to check performance on large data sets and improve where lacking. This is relevant for both regular grid N-D interpolators and QHull-based scattered N-D interpolators. For the latter ones, we additionally need to investigate if we can improve on the 32-bit integer limitations of the QHull library.

**Comment by mdhaber** [comment]
@lucascolley The CI section was

> Continuous integration currently covers 32/64-bit Windows, macOS on x86-64/arm,
32/64-bit Linux on x86, and Linux on aarch64 - as well as a range of versions
of our dependencies and building release quality wheels. Reliability of CI has
not been good recently (H1 2023), due to the large amount of configurations to
support and some CI jobs needing an overhaul. We aim to reduce build times by
removing the remaining distutils-based jobs when we drop that build system
and make the set of configurations in CI jobs more orthogonal.

It sounds like we should replace at least everything but the first sentence? With what? I know we want to add GPU tests, maybe `float32` tests? I don't personally have much to say about it, though, so I can't write that part.

Again, re: Accelerate: OK, we need to change it, but what is the new reason Accelerate is not the default for installations on a mac?

Re: substantial modifications to sections other then `stats`, `integrate`, `optimize`, `linalg`, and `differentiate`, I don't have much to say about them. Maybe those can be done in separate PRs by the appropriate people?

File: doc/source/dev/roadmap-detailed.rst (hunk 5 above)
Comment (ev-br) [reply doc/source/dev/roadmap-detailed.rst:176]: > That's the intended application, but I don't think it's useful for exploring the interface of such a interpolator.

why is that?

The example is cute, but to me it's quite firmly on the "best done at the call site" side. You just do exactly the same thing with any interpolator and that's it. And to my eyes the version with an calling a 1D interpolator twice or in a loop is more readable anyway.

Maybe there's a more "production"-grade example.

Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:176]: > why is that?

Your concern was about the interface, but `NumericalInverseHermite` does not have the interface that is anything like what we're talking about?

> The example is cute,

?
You wrote "Will complicate the interface at likely no perf gain. "
The example wasn't supposed to be cute... it was to demonstrate the proposed interface of the interpolator. 

> Maybe there's a more "production"-grade example.

How does that affect the interface of the interpolator? Wasn't the concern - that the interface of the interpolator would be too complicated?

---

Perhaps more context would help?

With the old infrastructure, we can create frozen distributions with array parameters:

```python3
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
X = stats.norm(loc=[[0], [1]], scale=[[1], [2]])
x = np.stack((np.linspace(-3, 3, 300), np.linspace(-4, 6, 300)))
plt.plot(x.T, X.pdf(x).T)
plt.show()
```
image

With the new infrastructure, we can do the same thing.

```python3
Y = stats.Normal(mu=[0, 1], sigma=[1, 2])
Y.plot()
plt.show()
```
image

Part of the purpose of `stats.sampling` is to represent distribution functions with splines so that it is faster to work with them, but it doesn't work if the frozen distribution has array shape parameters.
```
from scipy.stats import sampling
sampling.NumericalInverseHermite(X)
# TypeError: only length-1 arrays can be converted to Python scalars
```

If we wanted something like UNURAN that works when the distribution has array shape parameters, the interpolator needs to supporting different `x` coordinates for each set of parameters.

(I'll anticipate a valid criticism of the intended application - the `x`-grids might want to be different lengths. Agreed, but it is a separate issue from the interface of the interpolators.)

File: doc/source/dev/roadmap-detailed.rst (hunk 8 above)
Comment (rgommers) [reply doc/source/dev/roadmap-detailed.rst:82]: Here is how I'd rewrite this section:
```
SciPy owes a lot of its success to relying on wrapping well established
Fortran libraries (QUADPACK, FITPACK, ODRPACK, ODEPACK etc). The Fortran 77
that these libraries are written in is quite hard to maintain, and the use
of Fortran is problematic for many reasons, e.g., it makes our wheel builds
much harder to maintain, it has repeatedly been problematic for supporting
new platforms like macOS arm64 and Windows on Arm, and it is highly problematic
for Pyodide's SciPy support. Our goal is to remove all Fortran code from SciPy,
by replacing the functionality with code written in other languages. Progress
towards this goal is tracked in `gh-18566 <https://github.com/scipy/scipy/issues/18566>`__.
```

Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:82]: LGTM!

Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:82]: Besides `interpolate`, which is being handled in gh-22116, there is just one other mention of Fortran here, and that's in `sparse.linalg`. 

> - Fortran code not re-entrant (but we don't solve, maybe reuse from PyKrilov)

@ilayn, it looks like that conversion was already done? Should I just remove that line?

**Comment by DietBru** [comment]
> * signal:  is anything out of date here @DietBru ?

Two maintenance-related points come to mind:

1. Deprecate [_spectral_helper], [_triage_segments] and [_fft_helper]. 
Currently, [csd] is the only non-legacy function utilizing those internal functions. There already exists a [ShortTimeFFT]-based wrapper, for testing, namely [_csd_wrapper], to ensure that [ShortTimeFFT] provides a superset of functionality. This means that also the legacy functions [stft] and [spectrogram] will be deprecated.

2. Remove [To-Do List Wiki-Page].
It would be nice if the points on the [To-Do List Wiki-Page] could be updated and the still relevant entries could be converted into issues or be merged into this roadmap. 

[_spectral_helper]: https://github.com/scipy/scipy/blob/639bdb83ef66dfa49158edecafffb3eb463c9151/scipy/signal/_spectral_py.py#L1863
[_triage_segments]: https://github.com/scipy/scipy/blob/639bdb83ef66dfa49158edecafffb3eb463c9151/scipy/signal/_spectral_py.py#L2207
[_fft_helper]: https://github.com/scipy/scipy/blob/639bdb83ef66dfa49158edecafffb3eb463c9151/scipy/signal/_spectral_py.py#L2158
[stft]: https://github.com/scipy/scipy/blob/639bdb83ef66dfa49158edecafffb3eb463c9151/scipy/signal/_spectral_py.py#L1248
[spectrogram]: https://github.com/scipy/scipy/blob/639bdb83ef66dfa49158edecafffb3eb463c9151/scipy/signal/_spectral_py.py#L816
[csd]: https://github.com/scipy/scipy/blob/639bdb83ef66dfa49158edecafffb3eb463c9151/scipy/signal/_spectral_py.py#L660
[_csd_wrapper]: https://github.com/scipy/scipy/blob/639bdb83ef66dfa49158edecafffb3eb463c9151/scipy/signal/tests/_scipy_spectral_test_shim.py#L237
[ShortTimeFFT]: https://scipy.github.io/devdocs/reference/generated/scipy.signal.ShortTimeFFT.html
[To-Do List Wiki-Page]: https://github.com/scipy/scipy/wiki/%60scipy.signal%60-to%E2%80%90do-list

File: doc/source/dev/roadmap-detailed.rst (hunk 5 above)
Comment (ev-br) [reply doc/source/dev/roadmap-detailed.rst:181]: Thanks for the context, this is exactly what I meant to ask!
Let's continue in https://github.com/scipy/scipy/issues/22118

File: doc/source/dev/roadmap-detailed.rst (hunk 8 above)
Comment (ilayn) [reply doc/source/dev/roadmap-detailed.rst:82]: I think the needed parts are now quite outdated. LU decomposition situation has settled as lu_solve, lu_factor is for low-level users. We wrapped quite a number of LAPACK funcs since this was written and we don't get any requests anymore as we used to. Flapack part is also done/dormant. So I think these are all done except the batch issue. 

One needed item that comes to mind for linalg is, making SciPy, ILP64 compatible (LAPACK/BLAS using 64bit integers). It actually applies to all modules but predominantly hits linalg.

 > it looks like that conversion was already done? Should I just remove that line?

Yes, all Fortran rewrites are re-entrant and thread-safe.

**Comment by j-bowhay** [comment]
Should we look towards merging this? It seems to have reached consensus and is clearly a big improvement on the previous roadmap

**Comment by rgommers** [comment]
> Should we look towards merging this? It seems to have reached consensus and is clearly a big improvement on the previous roadmap

Seems okay to me once open comments are addressed. Leaving updates to content for other modules to a future PR is no problem I think.

**Comment by andyfaff** [comment]
The default MacOS wheels will use Accelerate for macOS >=14. It's only older OS versions that will use OpenBLAS.

File: doc/source/dev/roadmap-detailed.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply doc/source/dev/roadmap-detailed.rst:295]: Using this verbatim except the middle sentence; I would rather avoid the statements about being in good shape unless they are essential and hard to argue against.

**Comment by mdhaber** [comment]
I think I've addressed comments about the sections I aimed to updated. Most of https://github.com/scipy/scipy/pull/22101#pullrequestreview-2508692297, https://github.com/scipy/scipy/pull/22101#issuecomment-2550914451, https://github.com/scipy/scipy/pull/22101#discussion_r1887836382, and https://github.com/scipy/scipy/pull/22101#discussion_r1889772604 would be left to other PRs.

**Review by j-bowhay — APPROVED** [review]
Thanks all! Let's merge this then people are free to open follow ups to discuss changing other areas

**Comment by lucascolley** [comment]
follow-ups tracked in gh-22209, please add anything I missed - thanks all!
